File: ios/Classes/camera/data/CameraStatus.swift
//
//  StoreCameraPluginCameraStatus.swift
//  store_camera_plugin_camera
//
//  Created by 김경환 on 2020/10/22.
//

protocol CameraStatus: CameraPluginToMap {
    var type: CameraStatusType { get }
}

enum CameraStatusType: String {
    case BINDING = "BINDING"
    case ERROR = "ERROR"
}

struct CameraStatusBinding: CameraStatus {
    let type: CameraStatusType = CameraStatusType.BINDING
    let availablePosition: [String]
    let availableMotion: Bool
    let setting: CameraSetting
    
    func pluginToMap() -> [String : Any] {
        return [
            "type": type.rawValue,
            "availablePosition": availablePosition,
            "availableMotion": availableMotion,
            "setting": setting.pluginToMap()
        ]
    }
}

struct CameraStatusError: CameraStatus {
    let type: CameraStatusType = CameraStatusType.ERROR

    func pluginToMap() -> [String : Any] {
        return [
            "type": type.rawValue
        ]
    }
}

File: android/src/main/kotlin/io/storecamera/storecamera_photo/storecamera_plugin_camera/data/CameraStatus.kt
package io.storecamera.storecamera_photo.storecamera_plugin_camera.data

interface CameraStatus: PluginToMap {
    val type: CameraStatusType
}

enum class CameraStatusType(val value: String) {
    BINDING("BINDING"),
    ERROR("ERROR"),
}

class CameraStatusBinding(
        private val availablePosition: MutableList<String>,
        private val setting: CameraSetting): CameraStatus {
    override val type: CameraStatusType
        get() = CameraStatusType.BINDING
    
    override fun pluginToMap(): Map<String, *> = hashMapOf(
            "type" to type.value,
            "availablePosition" to availablePosition,
            "availableMotion" to true,
            "setting" to setting.pluginToMap()
    )
}

class CameraStatusError: CameraStatus {
    override val type: CameraStatusType
        get() = CameraStatusType.ERROR

    override fun pluginToMap(): Map<String, *> = hashMapOf(
            "type" to type.value
    )
}